### ISSUE TITLE
Pipe chain should start with a raw value.

### ISSUE THREAD

**Issue by seantanly** [rrrene/credo#32]
Thanks for Credo. :)

I'm not sure whether the above is a good refactoring opportunity.

For example, the code below triggers the above, and seems like a pretty common and valid style.

```
result = Stream.chunk(1..100, 10) |> Enum.take(3)

result = 
  if map[:foo] do
    map[:foo]
  else
    map[:default]
  end
  |> String.to_integer
````

**Comment by rrrene** [comment]
I don't disagree. Your examples show that we need exceptions from the rule.

The value of this check is when you do things like:

    String.downcase(params[:user_name])
    |> valid_username?()

I would argue that 

    params[:user_name]
    |> String.downcase
    |> valid_username?()

is so much clearer.

So what would you propose as "exceptional cases"? Ranges seem like a good candidate, like in your first example.

**Comment by seantanly** [comment]
I agree in your example, it can seem clearer. In fact, I'm habitually used to writing the pipe style as I code along, in anticipation that the output may be used in further transformation. Perhaps, I'm so used to looking at functions being called via pipe, that function calls with 1 less argument within the parentheses looks less cluttered/nicer to read.

E.g. I like
```
values = input_str |> String.split(~r/some_long_complex_ugly_looking_regex/, trim: true)
```

instead of 

```
values = String.split(input_str, ~r/some_long_complex_ugly_looking_regex/, trim: true)
```

However, the convention inside Elixir source code suggests otherwise. The pipe operator seems avoided when the output isn't further piped into another function.

In short, what I'm saying is this might be an area which boils down to personal taste. And it might be more noise than signal to suggest this refactoring.

**Comment by rrrene** [comment]
There is not much difference between our positions, I think.

We both agree that the "thing" going into the pipe should be clear. I see that in your initial examples, so I think rather than removing the check, I would like to search for sane defaults like "blocks and ranges can go into a pipe" with a possibility for configuration.

Will try to incooperate something into the `0.3.0` release :+1:

**Comment by chrismcg** [comment]
Thanks for Credo from me too. Another example from my code is:

```elixir
    Stream.resource(&stream_start/0, stream_data(api_call_fn), &stream_end/1)
    |> Stream.each(fn(data) -> send(caller, {message, data}) end)
    |> Stream.run
```

Where I have to introduce a variable to prevent the warning

```elixir
    stream = Stream.resource(&stream_start/0, stream_data(api_call_fn), &stream_end/1)
    stream
    |> Stream.each(fn(data) -> send(caller, {message, data}) end)
    |> Stream.run
```

Is there any plan to introduce something like `# credo:disable Refactor.PipeChain` ?

**Comment by rrrene** [comment]
@chrismcg There is, although I am still undecided on the format, since I personally dislike "magic comments". I would like to see a more "Elixiry" solution and I might just have had an idea :+1:

P.S. While I know this does solve the *actual* problem we are discussing, wouldn't this be the "Credo compatible solution"?

    &stream_start/0
    |> Stream.resource(stream_data(api_call_fn), &stream_end/1)
    |> Stream.each(fn(data) -> send(caller, {message, data}) end)
    |> Stream.run

**Comment by chrismcg** [comment]
Great to hear you have plans. In ruby I have rubocop as part of CI and it fails the build if it isn't happy and so being able to disable it for a particular section of code is necessary sometimes. At the moment I've Credo turned off on CI for my Elixir projects because I can't tell it "I know, I'll fix it later but it's not the most important thing to be doing right now"

I hadn't thought of your approach above. I don't think it's something I'd do though.

```elixir
foo
|> bar
|> baz
```

The code above tends to imply, in my mind at least, that `foo` here is being passed through each function, like `conn` in Phoenix or the stream in the example we've been talking about. I'm not passing the stream_start function through each call to stream so having it bare at the top of the pipe might be misleading compared to normal usage. Have to think about it more though.

**Comment by rrrene** [comment]
Yeah, I know what you mean with the `conn` example. I have had the same problem with actual production code that I thought to myself "is it really okay that the type changes constantly between pipes"? Is that maintainable code?

Contrived example:

    input_str                             # string with usernames
    |> String.split(",")                  # list of usernames
    |> Enum.map(&find_user_by_name/1)     # list of %User{} 
    |> Enum.filter(&(&1.email_enabled?))  # list of %User{} 
    |> Enum.each(&send_reminder/1)        # :ok

I am confident that we, as a community, will work things like these out over time. There's more open source Elixir code every day and more production code as well. Good times ahad :grin:

**Comment by chrismcg** [comment]
Agree there are good times ahead. I'm hoping later in the year I'll have more to contribute to Credo than examples on issues! In the meantime I still love chatting about this stuff though...

With the `input_str` example although you don't pass the string all the way through the rest of the pipe is concerned with transforming it. `&stream_start/0` is only used once in the `Stream.resource` call so I don't think it fits the metaphor of pipes transforming an input to an output.

**Comment by rrrene** [comment]
You're right, "taking and transforming data" is the purpose of pipes and the `&stream_start/0` example does not do that.

I will work on a method to exclude individual parts of code from certain checks :+1:

**Comment by rrrene** [comment]
@chrismcg Credo `v0.3.0` now includes support for a `@lint` attribute. See [CHANGELOG](https://github.com/rrrene/credo/blob/d6f039f4608b4ee73597a11debd8d52fcf63f804/CHANGELOG.md) for further details.

What do you think?

**Comment by chrismcg** [comment]
Looks great, thanks! Module attributes are a perfect solution for this.

**Comment by seantanly** [comment]
Thanks for @lint! And together with .credo.exs gives us full ability for customization.

**Comment by ryanwinchester** [comment]
Is there a way to disable the check if it is only 1 pipe and on a single line?

The example in OP is a good point for this

```elixir
result = Stream.chunk(1..100, 10) |> Enum.take(3)
```

I do this often like:

```elixir
bar = Foo.fun(foo) |> Bar.fun()
# ...
```

And honestly, I find this much better than 

```elixir
bar =
  foo
  |> Foo.fun()
  |> Bar.fun()

# ...
```

Heck, even...

```elixir
from(f in Foo, where: f.bar_id == ^bar_id) |> Repo.all()
```

Thoughts?

**Comment by hickscorp** [comment]
I would really like this to be extended to `case` statements too... For example:

```
    try do
      f.(data)
    rescue
      e -> {:error, {:ex, e, System.stacktrace()}}
    end
    |> case do
      {:ok, new_data} -> {:ok, {name, new_data}}
      err -> {err, state}
    end
    |> Tuple.insert_at(0, :reply)
```
looks really better than the non-pipe version. Am I missing something?

**Comment by whatyouhide** [comment]
@hickscorp the fact that that looks better is very much a matter of opinion. Personally, I think it's extremely confusing because of all the indentation and all the things going on. I'd much rather have things **named**:

```elixir
a =
  try do
    ...
  end

{reply, state} =
  case a do
    ...
  end

{:reply, reply, state}
```

**Comment by hickscorp** [comment]
@rrrene As I stated in the other issue, I agree to your comment, that it's a matter of taste. Therefore, a way of configuring that behaviour globally would be great.
